llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -883,8 +890,34 @@ bool AIEPreRASchedStrategy::isAvailableNode(SUnit &SU, SchedBoundary &Zone,
 
   // The node will likely cause a spill, only consider it schedule-able if
   // there is no pending node that can reduce the register pressure.
-  return findPressureReducer(WorstPC.getPSet(), Zone.Pending.elements(),
-                             BotRPT) == nullptr;
+  if (const SUnit *PendingPressureReducer = findPressureReducer(
+          WorstPC.getPSet(), Zone.Pending.elements(), BotRPT);
+      PendingPressureReducer && canBeDelayed(SU, *PendingPressureReducer)) {
+    LLVM_DEBUG(dbgs() << "** Delaying SU(" << SU.NodeNum << "): Waiting for SU("
+                      << PendingPressureReducer->NodeNum << ")\n");
+
+    // Keep track of PendingPressureReducer to avoid cycles of SUs
+    // delaying each other.
+    SUDelayerMap[SU.NodeNum] = PendingPressureReducer->NodeNum;
+    return false;
+  }
+
+  // Can't prove a pending SU will help reduce reg pressure, keep as available.
+  return true;
+}
+
+bool AIEPreRASchedStrategy::canBeDelayed(const SUnit &DelayedSU,
+                                         const SUnit &Delayer) const {
+  std::function<bool(unsigned)> Impl = [&](unsigned SUNum) {
+    if (SUNum == UnknownSUNum)
+      return false;
+    if (SUNum == DelayedSU.NodeNum)
+      return true;
+    return Impl(SUDelayerMap[SUNum]);
+  };
+  // If SU is delayed by another instruction that is eventually waiting on SU
+  // itself, do not keep delaying SU otherwise this creates an infinite loop.
+  return !Impl(Delayer.NodeNum);
```

Comment:
I can see it working for a cycle of two, but I'm not so sure about larger cycles. Could these occur?

Reply:
I could imagine something like:
- An instr decreases vec pressure but increases acc pressure
- An instr decreases acc pressure but increases gpr pressure
- An instr decreases gpr pressure but increases vec pressure

Reply:
paper, scissors, stone